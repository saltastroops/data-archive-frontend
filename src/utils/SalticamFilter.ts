export type SalticamFilter =
  | "SalticamFilter_Name"
  | "340-35"
  | "380-40"
  | "B-S1"
  | "CLR-S1"
  | "Halpha-S1"
  | "Hbn-S1"
  | "Hbw-S1"
  | "I-S1"
  | "R-S1"
  | "Sb-S1"
  | "SDSSg-S1"
  | "SDSSi-S1"
  | "SDSSr-S1"
  | "SDSSu-S1"
  | "SDSSz-S1"
  | "SR613-21"
  | "SR708-25"
  | "SR815-29"
  | "SR862-32"
  | "Su-S1"
  | "Sv-S1"
  | "Sy-S1"
  | "U-S1"
  | "V-S1";

export interface ISalticamFilter {
  name: SalticamFilter;
  descriptiveName: string;
}
