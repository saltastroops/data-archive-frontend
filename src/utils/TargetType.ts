export type TargetType = string;

export const GALAXY = "10";

export const ISM = "20";

export const SOLAR_SYSTEM_BODY = "30";

export const STAR = "40";
