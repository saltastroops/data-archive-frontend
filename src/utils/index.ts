export const toTitleCase = (s: string) => {
  s = s.replace(/_/g, " ");
  return s.replace(/\w\S*/g, (txt: string) => {
    return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase();
  });
};
