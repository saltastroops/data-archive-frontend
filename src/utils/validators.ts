import moment from "moment";

// TODO: Update once parse methods are available in util

/**
 * A method used to check if the given value(string) is a floating point number
 *
 * @param value:
 *     A string value to tested.
 * @return Boolean:
 *        True if the value is a float else false.
 */
export const isFloat = (value: string) => {
  const floatRegex = /^-?\d+(?:[.,]\d*?)?$/; // check optional sign, check decimal and check optional "." or "," followed by optional decimal
  if (!floatRegex.test(value)) {
    return false;
  }
  return !isNaN(parseFloat(value));
};

/**
 * This method check is given right ascension is within range and a correct time format (HH:MM:SS)
 *
 * @param time
 *    String representing Degrees arc minutes arc seconds
 * @return String/undefined
 *    String Error ifvtime can't be verified to be a aa valid time else nothing
 */
const raTimeError = (time: string) => {
  // regular expression to match required time format
  const re = /^(\d+):(\d+):(\d+)$/;

  const regs: any = time.match(re);
  if (regs) {
    // 24-hour value between 0 and 23
    if (regs[1] > 23) {
      return "Invalid value for hours: " + regs[1];
    }
    // minute value between 0 and 59
    if (regs[2] > 59) {
      return "Invalid value for minutes: " + regs[2];
    }
    // seconds value between 0 and 59
    if (regs[3] > 59) {
      return "Invalid value for seconds: " + regs[3];
    }
    return;
  } else {
    return "Right ascension should be in degrees or HH:MM:SS only";
  }
};

/**
 * This method check is given declination is within range
 *
 * @param degree
 *    String a float number
 * @return
 *    Error or nothing/undefined
 */
const validateDecDegrees = (degree: string) => {
  if (parseFloat(degree) < -90 || parseFloat(degree) > 90) {
    return "Declination should be between -90 and 90 degrees";
  }
  return;
};

/**
 * This method check is given declination is within range and a correct DMS
 *
 * @param dms
 *    String representing Degrees arc minutes arc seconds
 * @return String/undefined
 *    String Error if dms can't be verified to be a aa valid DMS else nothing
 */
const validateDecDms = (dms: string) => {
  const decs = dms.split(/[^0-9]+/).filter(d => d !== ""); // Split string at none numbers and remove white space and empty string
  if (!decs || decs.length !== 3) {
    // expecting 3 numbers separated by any thing else it should fail
    return "Declination should be in degrees or Degree:minutes:seconds";
  }
  if (-90 > parseFloat(decs[0]) || parseFloat(decs[0]) > 90) {
    // degress are between -90 and 90
    return "Declination should be between -90D 0M 0S and 90D 0M 0S degrees.\n";
  }
  if (
    (parseFloat(decs[0]) === 90 || parseFloat(decs[0]) === -90) &&
    (parseFloat(decs[1]) !== 0 || parseFloat(decs[2]) !== 0)
  ) {
    //
    return "Declination should be between -90D 0M 0S and 90D 0M 0S degrees.\n";
  }
  if (
    parseFloat(decs[1]) < 0 ||
    parseFloat(decs[1]) > 59 ||
    parseFloat(decs[2]) < 0 ||
    parseFloat(decs[2]) > 59
  ) {
    return "Declination minutes and seconds should be between 0 and 59";
  }
  return;
};

/**
 * This method test if string is a declination
 * It only accept two notation D"M'S and degrees else error
 *
 * @param dec:
 *     A string value to tested.
 * @return
 *    Error if string can not be a declination. else nothing
 */
export const validateDeclination = (dec: string) => {
  if (dec.trim() === "" || dec === null || dec === undefined) {
    return;
  }
  if (isFloat(dec)) {
    return validateDecDegrees(dec.trim());
  }

  let error;
  const decs = dec.split("..");
  if (decs.length === 1) {
    error = validateDecDms(dec.trim());
    if (error || error !== "") {
      return error;
    }
  }

  if (decs.length > 2) {
    return "Only two values of declination are allowed";
  }
  if (decs.length === 2) {
    // dec range
    const dec0 = decs[0].trim();
    const dec1 = decs[1].trim();
    if (isFloat(dec0) && isFloat(dec1)) {
      return validateDecDegrees(dec0) || validateDecDegrees(dec1);
    }
    return validateDecDms(dec0) || validateDecDms(dec1);
  }

  return;
};

export const validateName = () => {
  return "";
};

/**
 * This method test if string is a right ascension for range values need to separated by '..'
 * It only accept two notation HH:MM:SS and degrees else error
 *
 * @param ra:
 *     A string value to tested.
 * @return
 *    Error if string can not be a right ascension. else nothing
 */
export const validateRightAscension = (ra: string) => {
  let error = "";
  if (ra.trim() === "" || ra === null || ra === undefined) {
    return;
  }
  if (isFloat(ra)) {
    return 0 <= parseFloat(ra) && parseFloat(ra) <= 360
      ? undefined
      : "Declination should be between 0 and 360 degrees";
  }
  const ras = ra.split("..");
  if (ras.length > 2) {
    return 'Only two right ascension are permitted separated by ".."';
  }
  try {
    if (ras.length === 2) {
      const ra0 = ras[0].trim();
      const ra1 = ras[1].trim();
      if (isFloat(ra0)) {
        return 0 <= parseFloat(ra0) && parseFloat(ra0) <= 360
          ? undefined
          : "Declination should be between 0 and 360 degrees";
      }
      if (isFloat(ra1)) {
        return 0 <= parseFloat(ra1) && parseFloat(ra1) <= 360
          ? undefined
          : "Declination should be between 0 and 360 degrees";
      }
      error = raTimeError(ra1) || raTimeError(ra0) || "";
      if (error !== "") {
        return error;
      }
    }
    if (ras.length === 1) {
      error = raTimeError(ra) || "";
      if (error !== "") {
        return error;
      }
    }
  } catch (e) {
    return 'Invalid right ascension please use degrees or "HH:MM:SS"';
  }
  return;
};

/**
 * This method test if string can is a date(s)
 * Date range is separated by '..'
 *
 * @param date:
 *     A string value to tested.
 * @return :
 *    Error if string can not be a date. else nothing
 */
export const validateDate = (date: string) => {
  if (date === null || date === undefined || date.trim() === "") {
    return;
  }
  const re = /^(\d{4})-(\d{2})-(\d{2})$/;
  const regs: any = date.match(re);

  const dates = date.split("..");
  if (dates.length > 2) {
    return "Only two dates are permitted.";
  }
  if (dates.length === 2) {
    const day0 = dates[0].trim();
    const day1 = dates[1].trim();
    if (
      !day0.match(re) ||
      !day1.match(re) ||
      !moment(day0, "YYYY-MM-DD", true).isValid() ||
      !moment(day1, "YYYY-MM-DD", true).isValid()
    ) {
      return "You have an invalid date.";
    }
  }
  if (dates.length === 1) {
    if (
      !date.match(re) ||
      (!moment(date, "YYYY-MM-DD", true).isValid() || !regs)
    ) {
      return "You have an invalid date.";
    }
  }
  return;
};

export const validateSearchConeRadius = (radius: string) => {
  if (radius.trim() === "" || radius === null || radius === undefined) {
    return;
  }
  if (isFloat(radius) && parseFloat(radius) >= 0) {
    return;
  }
  return "Search radius should be a floating point number greater and zero.";
};

/**
 * Test if any of the given error groups (i.e. objects of error keys and
 * values) has a truthy value for any of its keys.
 *
 * Parameters
 * ----------
 * errorGroups:
 *     Objects of error keys and values.
 *
 * Returns
 * -------
 * isError:
 *     Whether any of the error groups has a truthy value.
 */
export const isError = (...errorGroups: object[]) => {
  return errorGroups.some((errorGroup: object) =>
    Object.keys(errorGroup).some(key => (errorGroup as any)[key])
  );
};
