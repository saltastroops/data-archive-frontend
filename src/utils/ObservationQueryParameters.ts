// TypeScript interfaces describing the state of the search form

import { SalticamFilter } from "./SalticamFilter";
import { TargetType } from "./TargetType";

/**
 * An interface describing the React state of the observation query form.
 *
 * This state is used for populating all search form fields (and error messages)
 * and is parsed into the GraphQL query for searching observations.
 *
 * Properties:
 * -----------
 * general:
 *     General information about the observation.
 * target:
 *     Target details.
 * telescope:
 *     Telescope (and instrument) details.
 */
// TODO: Remove loading, replace with IObservationQueryParameters
export interface ISearchFormState {
  general: IGeneral;
  target: ITarget;
  telescope?: ITelescope;
  loading: boolean;
}

/**
 * The available calibration types.
 */
export type CalibrationType = "arc" | "bias" | "flat" | "standard";

/**
 * An interface for query parameters related to general information.
 *
 * Properties:
 * -----------
 * errors:
 *     Error messages.
 * observationNight:
 *     Date of the night when the observation was taken. This
 *     refers to the date whe the night started.
 * proposalCode:
 *     Identifier of the proposal to which the observation belongs.
 * proposalTitle:
 *     Title of the proposal to which the observation belongs.
 * calibrations:
 *     Calibration types to include in the search results
 *
 */
export interface IGeneralErrors {
  observationNight?: string;
  principalInvestigator?: string;
  proposalCode?: string;
  proposalTitle?: string;
  calibrations?: string;
}
export interface IGeneral {
  errors: IGeneralErrors;
  observationNight?: string;
  principalInvestigator?: string;
  proposalCode?: string;
  proposalTitle?: string;
  calibrations: Set<CalibrationType>;
}

/**
 * The available target name resolvers.
 */
export type TargetResolver = "Simbad" | "NED" | "VizieR";

/**
 * The available units for the search cone radius.
 */
export type SearchConeRadiusUnits = "arcseconds" | "arcminutes" | "degrees";

/**
 * An interface for query parameters related to target details.
 *
 * Properties:
 * -----------
 * declination:
 *     Declination.
 * errors:
 *     Errors.
 * name:
 *     Target name.
 * resolver:
 *     Resolver to use for resolving the target name to a position.
 * rightAscension:
 *     Right ascension.
 * searchConeRadius:
 *     Radius of the search cone radius, in the units given by
 *     searchConeRadiusUnits.
 * searchConeRadiusUnits:
 *     Units of the search cone radius given by searchConeRadius.
 *
 */
export interface ITarget {
  declination?: string;
  errors: ITargetErrors;
  loading?: string;
  name?: string;
  resolver: TargetResolver;
  rightAscension?: string;
  searchConeRadius?: string;
  searchConeRadiusUnits: SearchConeRadiusUnits;
  targetTypes: Set<TargetType>;
}

export interface ITargetErrors {
  name?: string;
  declination?: string;
  resolver?: string;
  rightAscension?: string;
  searchConeRadius?: string;
  searchConeRadiusUnits?: string;
  targetTypes?: string;
}

// TELESCOPES

/**
 * The supported telescope names.
 */
export type TelescopeName = "Lesedi" | "SALT" | "1.9 m";

/**
 * An interface for query parameters related to a telescope.
 *
 * This serves as a placeholder for more concrete interfaces such as ISALT.
 *
 * Properties:
 * -----------
 * name:
 *     Telescope name.
 */
export interface ITelescope {
  name?: TelescopeName;
}

/**
 * An interface for query parameters related to SALT.
 *
 * Properties:
 * -----------
 * instrument:
 *     A SALT instrument.
 * name:
 *     The string "SALT".
 */
export interface ISALT extends ITelescope {
  instrument: IInstrument;
  name: "SALT";
}

/**
 * An interface for query parameters related to Lesedi.
 *
 * Properties:
 * -----------
 * instrument:
 *     An instrument used on Lesedi.
 * name:
 *     The string "Lesedi".
 */
export interface ILesedi extends ITelescope {
  instrument: IInstrument;
  name: "Lesedi";
}

/**
 * An interface for query parameters related to the 1.9 m Telescope.
 *
 * Properties:
 * -----------
 * instrument:
 *     An instrument used on the 1./9 m Telescope.
 * name:
 *     The string "1.9 m".
 */
export interface IOneDotNineM extends ITelescope {
  instrument: IInstrument;
  name: "1.9 m";
}

// INSTRUMENTS

/**
 * The available instrument names.
 */
export type InstrumentName =
  | "Salticam"
  | "RSS"
  | "HRS"
  | "BVIT"
  | "SpUpNIC"
  | "SHOC"
  | "HIPPO";

/**
 * An interface for query parameters related to an instrument.
 *
 * This serves as a placeholder for more concrete interfaces such as ISalticam.
 *
 * Properties:
 * -----------
 * name:
 *     Telescope name.
 */
export interface IInstrument {
  name: InstrumentName;
}

/**
 * An interface for query parameters related to Salticam.
 *
 * Properties:
 * -----------
 * detectorMode:
 *     Detector mode.
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 * name:
 *     The string "Salticam".
 */
export interface ISalticam extends IInstrument {
  detectorMode?: "Normal" | "Slot Mode";
  errors: {
    detectorMode?: string;
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: SalticamFilter;
  name: "Salticam";
}

/**
 * An interface for query parameters related to RSS.
 *
 * Properties:
 * -----------
 * detectorMode:
 *     Detector mode.
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 * mode:
 *     Instrument mode specific details.
 * name:
 *     The string "RSS".
 */
export interface IRSS extends IInstrument {
  detectorMode?: RSSDetectorMode;
  errors: {
    detectorMode?: string;
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
  modes?: IRSSModes;
  name: "RSS";
}

export interface IRSSModes {
  errors: {
    fabryPerotMode?: string;
    grating?: string;
    polarimetryModes?: string;
  };
  fabryPerotMode?: RSSFabryPerotMode;
  grating?: RSSGrating;
  names: Set<RSSInstrumentMode>;
  polarimetryModes?: Set<RSSPolarimetryMode>;
}

export type RSSDetectorMode = "Normal" | "Slot Mode" | "";

export type RSSGrating =
  | "Open"
  | "pg0300"
  | "pg0900"
  | "pg1300"
  | "pg1800"
  | "pg2300"
  | "pg3000"
  | "";

export type RSSInstrumentMode =
  | "Fabry Perot"
  | "FP polarimetry"
  | "Imaging"
  | "Polarimetric imaging"
  | "MOS"
  | "MOS polarimetry"
  | "Spectropolarimetry"
  | "Spectroscopy";

export type RSSFabryPerotMode = "HR" | "LR" | "MR" | "TF" | "";

export type RSSPolarimetryMode =
  | "All Stokes"
  | "Circular"
  | "Linear"
  | "Linear Hi"
  | "";

/**
 * An interface for query parameters related to HRS.
 *
 * Properties:
 * -----------
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * mode:
 *     Resolution mode.
 * name:
 *     The string "HRS".
 */
export interface IHRS extends IInstrument {
  errors: {
    exposureTime?: string;
    mode?: string;
  };
  exposureTime?: string;
  mode?: HRSMode;
  name: "HRS";
}

export type HRSMode =
  | "High Resolution"
  | "High Stability"
  | "Int Cal Fibre"
  | "Low Resolution"
  | "Medium Resolution";

export interface IBVIT extends IInstrument {
  errors: {};
  filter?: BVITFilter;
  mode?: "Imaging" | "Streaming";
  name: "BVIT";
}

export type BVITFilter = "B" | "H-alpha" | "Open" | "R" | "U" | "V";

/**
 * An interface for query parameters related to HIPPO.
 *
 * Properties:
 * -----------
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 * name:
 *     The string "HIPPO".
 */
export interface IHIPPO extends IInstrument {
  errors: {
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
  name: "HIPPO";
}

/**
 * An interface for query parameters related to SHOC.
 *
 * Properties:
 * -----------
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 * name:
 *     The string "SHOC".
 */
export interface ISHOC extends IInstrument {
  errors: {
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
  name: "SHOC";
}

/**
 * An interface for query parameters related to SpUpNIC.
 *
 * Properties:
 * -----------
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 * name:
 *     The string "SpUpNIC".
 */
export interface ISpUpNIC extends IInstrument {
  errors: {
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
  name: "SpUpNIC";
}

export interface IFile {
  id: string;
  filename: string;
  name: string;
  dataType: string;
  isReduced: boolean;
  targetName: string;
  rightAscension: number;
  declination: number;
  observationNight: string;
  category: string;
  telescope: string;
  instrument: string;
  observationId?: string;
  url?: string;
}
export interface IObservation {
  id: string;
  name: string | "Unknown Observation";
  proposal?: string;
  telescope: string;
  startTime: string;
  files: IFile[];
}
