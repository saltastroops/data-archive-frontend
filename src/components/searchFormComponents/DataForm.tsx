import * as React from "react";
import styled from "styled-components";
import { IGeneral, ProductType } from "../../utils/ObservationQueryParameters";
import {
  SingleColumnGrid,
  Span,
  SubGrid,
  SubGrid5
} from "../basicComponents/Grids";

const LargeCheckbox = styled.input.attrs({
  className: "checkbox",
  type: "checkbox"
})`
  && {
    width: 18px;
    height: 18px;
  }
`;

interface IDataFormProps {
  general: IGeneral;
  onChange: (value: any) => void;
}

/**
 * A form for choosing what data to include in an observation search.
 */
class DataForm extends React.Component<IDataFormProps, {}> {
  // Add or remove the calibration type corresponding to the clicked checkbox
  changeCheckbox = (e: React.ChangeEvent<HTMLInputElement>) => {
    const name = e.target.name as any;
    let { rejected } = this.props.general;
    const updated = new Set<ProductType>(this.props.general.productTypes);
    if (e.target.checked) {
      if (name === "rejected") {
        rejected = true;
      } else {
        updated.add(name);
      }
    } else {
      if (name === "rejected") {
        rejected = false;
      } else {
        updated.delete(name);
      }
    }
    this.props.onChange({
      ...this.props.general,
      productTypes: updated,
      rejected
    });
  };

  render() {
    const { productTypes, rejected } = this.props.general;
    return (
      <>
        <SingleColumnGrid>
          <SubGrid>
            <h5 className={"title is-5"}>Data category</h5>
          </SubGrid>
        </SingleColumnGrid>
        <SingleColumnGrid>
          <SubGrid5>
            <label>
              <Span>
                <LargeCheckbox
                  id="science-checkbox"
                  checked={productTypes.has("Science")}
                  data-test="science-checkbox"
                  onChange={this.changeCheckbox}
                  name="science"
                />
              </Span>
              <Span>Science</Span>
            </label>
            <label>
              <Span>
                <LargeCheckbox
                  id="arcs-checkbox"
                  checked={productTypes.has("Arc")}
                  data-test="arcs-checkbox"
                  onChange={this.changeCheckbox}
                  name="arc"
                />
              </Span>
              <Span>Arcs</Span>
            </label>
            <label>
              <Span>
                <LargeCheckbox
                  id="biases-checkbox"
                  checked={productTypes.has("Bias")}
                  data-test="biases-checkbox"
                  onChange={this.changeCheckbox}
                  name="bias"
                />
              </Span>
              <Span>Biases</Span>
            </label>
            <label>
              <Span>
                <LargeCheckbox
                  id="flats-checkbox"
                  checked={productTypes.has("Flat")}
                  data-test="flats-checkbox"
                  onChange={this.changeCheckbox}
                  name="flat"
                />
              </Span>
              <Span>Flats</Span>
            </label>
            <label>
              <Span>
                <LargeCheckbox
                  id="standards-checkbox"
                  checked={productTypes.has("Standard")}
                  data-test="standards-checkbox"
                  onChange={this.changeCheckbox}
                  name="standard"
                />
              </Span>
              <Span>Standards</Span>
            </label>
          </SubGrid5>
        </SingleColumnGrid>

        <SingleColumnGrid>
          <SubGrid>
            <h5 className={"title is-5"}>Data status</h5>
          </SubGrid>
        </SingleColumnGrid>
        <SingleColumnGrid>
          <SubGrid>
            <label>
              <Span>
                <LargeCheckbox
                  id="rejected-checkbox"
                  checked={rejected}
                  data-test="rejected-checkbox"
                  onChange={this.changeCheckbox}
                  name="rejected"
                />
              </Span>
              <Span>Rejected</Span>
            </label>
          </SubGrid>
        </SingleColumnGrid>
      </>
    );
  }
}
export default DataForm;
