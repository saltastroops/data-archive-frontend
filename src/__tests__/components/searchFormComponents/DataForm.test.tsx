import { mount, ReactWrapper, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import DataForm from "../../../components/searchFormComponents/DataForm";
import { ProductType } from "../../../utils/ObservationQueryParameters";

const onChange = jest.fn();

describe("Data Form", () => {
  let wrapper: ReactWrapper;

  beforeEach(() => {
    wrapper = mount(
      <DataForm
        general={{ productTypes: new Set<ProductType>(), errors: {} }}
        onChange={onChange}
      />
    );
  });

  it("should render", () => {
    expect(wrapper).toBeDefined();
  });

  it("should render correctly", () => {
    expect(
      toJson(
        shallow(
          <DataForm
            general={{
              errors: {},
              productTypes: new Set<ProductType>(["arc"])
            }}
            onChange={onChange}
          />
        )
      )
    ).toMatchSnapshot();
    expect(
      toJson(
        shallow(
          <DataForm
            general={{
              errors: {},
              productTypes: new Set<ProductType>(["Arc", "Standard"])
            }}
            onChange={onChange}
          />
        )
      )
    ).toMatchSnapshot();
    expect(
      toJson(
        shallow(
          <DataForm
            general={{
              errors: {},
              productTypes: new Set<ProductType>([
                "Arc",
                "Bias",
                "Flat",
                "Standard"
              ])
            }}
            onChange={onChange}
          />
        )
      )
    ).toMatchSnapshot();
  });

  it("should call onChange with the correct values", () => {
    // Select flats...
    const flats = wrapper.find('[data-test="flats-checkbox"] input');
    flats.simulate("change", {
      target: { checked: true, name: "Flat" }
    });
    expect(onChange).toBeCalledWith({
      errors: {},
      productTypes: new Set<ProductType>(["Flat"])
    });

    // ... and unselect them again
    flats.simulate("change", {
      target: { checked: false, name: "Flat" }
    });
    expect(onChange).toBeCalledWith({
      errors: {},
      productTypes: new Set<ProductType>([])
    });

    // select arcs
    const arcs = wrapper.find('[data-test="arcs-checkbox"] input');
    arcs.simulate("change", {
      target: { checked: true, name: "Arc" }
    });
    expect(onChange).toBeCalledWith({
      errors: {},
      productTypes: new Set<ProductType>(["Arc"])
    });
  });
});
