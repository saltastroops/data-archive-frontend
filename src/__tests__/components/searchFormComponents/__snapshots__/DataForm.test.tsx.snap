// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Form should render correctly 1`] = `
<Fragment>
  <styled.div>
    <styled.div>
      <h5
        className="title is-5"
      >
        Data category
      </h5>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <label>
        <styled.span>
          <styled.input
            checked={false}
            data-test="science-checkbox"
            id="science-checkbox"
            name="science"
            onChange={[Function]}
          />
        </styled.span>
        <styled.span>
          Science
        </styled.span>
      </label>
      <label>
        <styled.span>
          <styled.input
            checked={false}
            data-test="arcs-checkbox"
            id="arcs-checkbox"
            name="arc"
            onChange={[Function]}
          />
        </styled.span>
        <styled.span>
          Arcs
        </styled.span>
      </label>
      <label>
        <styled.span>
          <styled.input
            checked={false}
            data-test="biases-checkbox"
            id="biases-checkbox"
            name="bias"
            onChange={[Function]}
          />
        </styled.span>
        <styled.span>
          Biases
        </styled.span>
      </label>
      <label>
        <styled.span>
          <styled.input
            checked={false}
            data-test="flats-checkbox"
            id="flats-checkbox"
            name="flat"
            onChange={[Function]}
          />
        </styled.span>
        <styled.span>
          Flats
        </styled.span>
      </label>
      <label>
        <styled.span>
          <styled.input
            checked={false}
            data-test="standards-checkbox"
            id="standards-checkbox"
            name="standard"
            onChange={[Function]}
          />
        </styled.span>
        <styled.span>
          Standards
        </styled.span>
      </label>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <h5
        className="title is-5"
      >
        Data status
      </h5>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <label>
        <styled.span>
          <styled.input
            checked={false}
            data-test="rejected-checkbox"
            id="rejected-checkbox"
            name="rejected"
            onChange={[Function]}
          />
        </styled.span>
        <styled.span>
          Rejected
        </styled.span>
      </label>
    </styled.div>
  </styled.div>
</Fragment>
`;

exports[`Data Form should render correctly 2`] = `
<Fragment>
  <styled.div>
    <styled.div>
      <h5
        className="title is-5"
      >
        Data category
      </h5>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <label>
        <styled.span>
          <styled.input
            checked={false}
            data-test="science-checkbox"
            id="science-checkbox"
            name="science"
            onChange={[Function]}
          />
        </styled.span>
        <styled.span>
          Science
        </styled.span>
      </label>
      <label>
        <styled.span>
          <styled.input
            checked={false}
            data-test="arcs-checkbox"
            id="arcs-checkbox"
            name="arc"
            onChange={[Function]}
          />
        </styled.span>
        <styled.span>
          Arcs
        </styled.span>
      </label>
      <label>
        <styled.span>
          <styled.input
            checked={false}
            data-test="biases-checkbox"
            id="biases-checkbox"
            name="bias"
            onChange={[Function]}
          />
        </styled.span>
        <styled.span>
          Biases
        </styled.span>
      </label>
      <label>
        <styled.span>
          <styled.input
            checked={false}
            data-test="flats-checkbox"
            id="flats-checkbox"
            name="flat"
            onChange={[Function]}
          />
        </styled.span>
        <styled.span>
          Flats
        </styled.span>
      </label>
      <label>
        <styled.span>
          <styled.input
            checked={false}
            data-test="standards-checkbox"
            id="standards-checkbox"
            name="standard"
            onChange={[Function]}
          />
        </styled.span>
        <styled.span>
          Standards
        </styled.span>
      </label>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <h5
        className="title is-5"
      >
        Data status
      </h5>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <label>
        <styled.span>
          <styled.input
            checked={false}
            data-test="rejected-checkbox"
            id="rejected-checkbox"
            name="rejected"
            onChange={[Function]}
          />
        </styled.span>
        <styled.span>
          Rejected
        </styled.span>
      </label>
    </styled.div>
  </styled.div>
</Fragment>
`;

exports[`Data Form should render correctly 3`] = `
<Fragment>
  <styled.div>
    <styled.div>
      <h5
        className="title is-5"
      >
        Data category
      </h5>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <label>
        <styled.span>
          <styled.input
            checked={false}
            data-test="science-checkbox"
            id="science-checkbox"
            name="science"
            onChange={[Function]}
          />
        </styled.span>
        <styled.span>
          Science
        </styled.span>
      </label>
      <label>
        <styled.span>
          <styled.input
            checked={false}
            data-test="arcs-checkbox"
            id="arcs-checkbox"
            name="arc"
            onChange={[Function]}
          />
        </styled.span>
        <styled.span>
          Arcs
        </styled.span>
      </label>
      <label>
        <styled.span>
          <styled.input
            checked={false}
            data-test="biases-checkbox"
            id="biases-checkbox"
            name="bias"
            onChange={[Function]}
          />
        </styled.span>
        <styled.span>
          Biases
        </styled.span>
      </label>
      <label>
        <styled.span>
          <styled.input
            checked={false}
            data-test="flats-checkbox"
            id="flats-checkbox"
            name="flat"
            onChange={[Function]}
          />
        </styled.span>
        <styled.span>
          Flats
        </styled.span>
      </label>
      <label>
        <styled.span>
          <styled.input
            checked={false}
            data-test="standards-checkbox"
            id="standards-checkbox"
            name="standard"
            onChange={[Function]}
          />
        </styled.span>
        <styled.span>
          Standards
        </styled.span>
      </label>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <h5
        className="title is-5"
      >
        Data status
      </h5>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <label>
        <styled.span>
          <styled.input
            checked={false}
            data-test="rejected-checkbox"
            id="rejected-checkbox"
            name="rejected"
            onChange={[Function]}
          />
        </styled.span>
        <styled.span>
          Rejected
        </styled.span>
      </label>
    </styled.div>
  </styled.div>
</Fragment>
`;
